File: go/provenance.go
/* Copyright (c) 2025 Richard Rodger, MIT License */


package aontu

import (
	"sort"
	"strings"
)

const (
	WhyLiteral = "literal"
	WhySpread  = "spread"
	WhyRef     = "ref"
	WhyPref    = "pref"
)

// WhySite is the G2 site object, minus its data/schema role: a
// contribution's role is its own, and a `why` run has one document.
type WhySite struct {
	Col  int    `json:"col"`
	File string `json:"file"`
	Len int `json:"len"`
	Row int `json:"row"`
}

type WhyConjunct struct {
	Canon string `json:"canon"`
	Rank *int    `json:"rank,omitempty"`
	Role string  `json:"role"`
	Site WhySite `json:"site"`
	Src string `json:"src"`
}

type WhyRecord struct {
	Conjuncts []WhyConjunct `json:"conjuncts"`
	Path      string        `json:"path"`
	Value     string        `json:"value"`
}

type whyContribution struct {
	WhyConjunct
	val Val
}

type whyPathRecord struct {
	conjuncts []whyContribution
	// Ids of values PRODUCED by a meet at this path: an operand among
	// them is an intermediate result, not a source contribution.
	made map[Val]bool
	seen map[Val]bool
}

// Provenance is the recorder itself. Mirrors the class in
// ts/src/provenance.ts.
type Provenance struct {
	paths map[string]*whyPathRecord
	// The entry source, for turning a byte offset into row and column.
	src string
	texts map[string]string
}

func newProvenance(src string, texts map[string]string) *Provenance {
	return &Provenance{
		paths: map[string]*whyPathRecord{},
		src:   src,
		texts: texts,
	}
}

func (p *Provenance) writtenFrom(v Val) {
	if nil == v || v.written() {
		return
	}
	v.setWritten()
	for _, k := range whyKids(v) {
		p.writtenFrom(k)
	}
	// OUTERMOST WINS: the walk is top-down, so a value already pointed
	// at a container is inside that one and this one, and the answer the
	// author wants is the whole written statement. See base.finner.
	for _, k := range samePathKids(v) {
		if nil == k.innerOf() {
			k.setInnerOf(v)
		}
	}
}

func samePathKids(v Val) []Val {
	switch b := v.(type) {
	case *DisjunctVal:
		return b.peg
	case *PrefVal:
		if nil == b.peg {
			return nil
		}
		return []Val{b.peg}
	case *FuncVal:
		return b.peg
	case *PlusOpVal:
		return b.peg
	}
	return nil
}

// whyKids is the structural walk both the written stamp and the
// inside-set share: bag children and spread, junction members, func
// arguments, a preference's value.
func whyKids(v Val) []Val {
	switch b := v.(type) {
	case *MapVal:
		out := make([]Val, 0, len(b.keys)+1)
		for _, k := range b.keys {
			out = append(out, b.peg[k])
		}
		if nil != b.spread {
			out = append(out, b.spread)
		}
		return out
	case *ListVal:
		out := append([]Val{}, b.peg...)
		if nil != b.spread {
			out = append(out, b.spread)
		}
		return out
	case *ConjunctVal:
		return b.peg
	case *DisjunctVal:
		return b.peg
	case *PrefVal:
		return []Val{b.peg}
	case *FuncVal:
		return b.peg
	case *PlusOpVal:
		return b.peg
	}
	return nil
}

// whyRoleRank orders the roles by how much they tell the reader, for
// the deduplication in `at`: a role that names HOW a value reached the
// path says more than "written here".
func whyRoleRank(role string) int {
	switch role {
	case WhySpread:
		return 0
	case WhyRef:
		return 1
	case WhyPref:
		return 2
	}
	return 3
}

func (p *Provenance) whyRole(v Val) string {
	if v.fromSpread() {
		return WhySpread
	}
	switch v.(type) {
	case *RefVal:
		return WhyRef
	case *PrefVal:
		return WhyPref
	}
	return WhyLiteral
}

// record is one meet. Both operands are candidate contributions; the
// result is remembered so a later meet does not mistake it for a
// source.
func (p *Provenance) record(path []string, a, b, out Val) {
	key := strings.Join(path, ".")
	rec, ok := p.paths[key]
	if !ok {
		rec = &whyPathRecord{
			made: map[Val]bool{},
			seen: map[Val]bool{},
		}
		p.paths[key] = rec
	}

	p.contribute(rec, a)
	p.contribute(rec, b)

	if nil != out && out != a && out != b {
		rec.made[out] = true
	}
}

func (p *Provenance) contribute(rec *whyPathRecord, v Val) {
	if nil == v || isTop(v) || v.Nil() ||
		rec.made[v] || rec.seen[v] {
		return
	}
	// Not the author's: see base.fwrt.
	if !v.written() && !v.fromSpread() {
		return
	}

	// PART OF a written value is not a value beside it: report the
	// whole statement the author wrote, whichever piece of it the
	// fixpoint happened to meet here. See base.finner.
	outer := v
	for up := outer.innerOf(); nil != up; up = outer.innerOf() {
		outer = up
	}
	if outer != v {
		p.contribute(rec, outer)
		return
	}
	// A CONJUNCT is not one contribution, it is the statement that
	// several must all hold — duplicate keys merged at parse, an
	// explicit `a & b`. Its own site is nowhere (the merge has no
	// source position), while its terms each have one, which is what
	// the author needs to be shown.
	if cj, isc := v.(*ConjunctVal); isc {
		rec.seen[v] = true
		for _, term := range cj.peg {
			p.contribute(rec, term)
		}
		return
	}

	rec.seen[v] = true

	row, col := -1, -1
	if 0 <= v.pos() {
		text, have := p.src, true
		if file := v.srcurl(); "" != file {
			text, have = p.texts[file]
		}
		if have {
			row, col = rowCol(text, v.pos())
		}
	}
	c := whyContribution{
		WhyConjunct: WhyConjunct{
			Canon: v.Canon(),
			Role:  p.whyRole(v),
			Site: WhySite{
				Col: col, File: v.srcurl(), Len: v.srclen(), Row: row,
			},
			Src: v.srctext(),
		},
		val: v,
	}
	if pv, isp := v.(*PrefVal); isp {
		rank := pv.rank
		c.Rank = &rank
	}
	rec.conjuncts = append(rec.conjuncts, c)
}

func (p *Provenance) stands(path []string, v Val) {
	rec, ok := p.paths[strings.Join(path, ".")]
	if ok && 0 < len(rec.conjuncts) {
		return
	}
	p.record(path, v, nil, nil)
}

func (p *Provenance) at(path []string) []WhyConjunct {
	rec, ok := p.paths[strings.Join(path, ".")]
	if !ok {
		return []WhyConjunct{}
	}
	sorted := append([]whyContribution{}, rec.conjuncts...)
	sort.SliceStable(sorted, func(i, j int) bool {
		a, b := sorted[i].WhyConjunct, sorted[j].WhyConjunct
		if a.Site.File != b.Site.File {
			return a.Site.File < b.Site.File
		}
		if a.Site.Row != b.Site.Row {
			return a.Site.Row < b.Site.Row
		}
		if a.Site.Col != b.Site.Col {
			return a.Site.Col < b.Site.Col
		}
		return a.Canon < b.Canon
	})

	kept := []*WhyConjunct{}
	shown := map[string]*WhyConjunct{}
	for _, c := range sorted {
		one := c.WhyConjunct
		if 0 > c.Site.Row {
			kept = append(kept, &one)
			continue
		}
		key := strings.Join([]string{
			c.Src, c.Site.File,
			itoa(c.Site.Row), itoa(c.Site.Col), itoa(c.Site.Len),
		}, "\x00")
		had, seen := shown[key]
		if !seen {
			kept = append(kept, &one)
			shown[key] = &one
			continue
		}
		if whyRoleRank(c.Role) < whyRoleRank(had.Role) {
			had.Role = c.Role
		}
	}
	out := make([]WhyConjunct, 0, len(kept))
	for _, c := range kept {
		out = append(out, *c)
	}
	return out
}

File: ts/src/provenance.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


export type WhyRole = 'literal' | 'spread' | 'ref' | 'pref'

// The G2 site object, minus its data/schema role: a contribution's
// role is its own (above), and a `why` run has one document.
export type WhySite = {
  col: number
  file: string
  len: number
  row: number
}

export type WhyConjunct = {
  canon: string
  role: WhyRole
  site: WhySite
  src: string
  rank?: number
}

export type WhyRecord = {
  conjuncts: WhyConjunct[]
  path: string
  value: string
}


// Set on a spread template's per-key clone, at the one place a spread
// is applied (MapVal/ListVal.unify). Nothing else reads it.
export const FROM_SPREAD = '_fromSpread'


export const WRITTEN = '_written'


export const INNER_OF = '_innerOf'


export function markSpread(v: any, seen?: Set<any>): void {
  const marked = seen ?? new Set<any>()
  if (null == v || true !== v.isVal || marked.has(v)) {
    return
  }
  marked.add(v)
  v[FROM_SPREAD] = true
  if (true === v.isMap && null != v.peg) {
    for (const k of Object.keys(v.peg)) {
      markSpread(v.peg[k], marked)
    }
  }
  else if (true === v.isList && null != v.peg) {
    for (const k of Object.keys(v.peg)) {
      markSpread(v.peg[k], marked)
    }
  }
  else if (Array.isArray(v.peg)) {
    // A junction, a func's arguments, an op's operands: every one of
    // them can hold the value that reaches the destination.
    for (const m of v.peg) {
      markSpread(m, marked)
    }
  }
  else if (true === v.isPref) {
    markSpread(v.peg, marked)
  }
}


// The children that stand at the SAME path as v. See INNER_OF.
function samePathKids(v: any): any[] {
  if (true === v.isMap || true === v.isList || true === v.isConjunct) {
    return []
  }
  if (true === v.isPref) {
    return null != v.peg && true === v.peg.isVal ? [v.peg] : []
  }
  return Array.isArray(v.peg)
    ? v.peg.filter((k: any) => null != k && true === k.isVal) : []
}


function cmpSite(a: Contribution, b: Contribution): number {
  return a.site.file.localeCompare(b.site.file) ||
    a.site.row - b.site.row ||
    a.site.col - b.site.col ||
    a.canon.localeCompare(b.canon)
}


function roleOf(v: any): WhyRole {
  if (true === v[FROM_SPREAD]) {
    return 'spread'
  }
  if (true === v.isRef) {
    return 'ref'
  }
  if (true === v.isPref) {
    return 'pref'
  }
  return 'literal'
}


type Contribution = WhyConjunct & {
  id: number
}

type PathRecord = {
  conjuncts: Contribution[]
  // Ids of values PRODUCED by a meet at this path: an operand among
  // them is an intermediate result, not a source contribution.
  made: Set<number>
  seen: Set<number>
}


export class Provenance {
  paths: Map<string, PathRecord> = new Map()

  containers: Map<number, any> = new Map()

  writtenFrom(v: any): void {
    if (null == v || true !== v.isVal || true === v[WRITTEN]) {
      return
    }
    v[WRITTEN] = true
    const kids: any[] =
      (true === v.isMap || true === v.isList) && null != v.peg
        ? Object.keys(v.peg).map((k) => v.peg[k])
        : Array.isArray(v.peg) ? v.peg
          : null != v.peg && true === v.peg.isVal ? [v.peg]
            : []
    for (const k of kids) {
      this.writtenFrom(k)
    }
    // OUTERMOST WINS: the walk is top-down, so a value already pointed
    // at a container is inside that one and this one, and the answer
    // the author wants is the whole written statement.
    for (const k of samePathKids(v)) {
      if (null == k[INNER_OF]) {
        k[INNER_OF] = v.id
        this.containers.set(v.id, v)
      }
    }
    if (null != v.spread?.cj) {
      this.writtenFrom(v.spread.cj)
    }
  }

  // One meet. Both operands are candidate contributions; the result is
  // remembered so a later meet does not mistake it for a source.
  record(path: string[], a: any, b: any, out: any): void {
    const key = path.join('.')
    let rec = this.paths.get(key)
    if (null == rec) {
      rec = {
        conjuncts: [], made: new Set(), seen: new Set(),
      }
      this.paths.set(key, rec)
    }

    this.contribute(rec, a)
    this.contribute(rec, b)

    if (null != out && true === out.isVal && out !== a && out !== b) {
      rec.made.add(out.id)
    }
  }

  private contribute(rec: PathRecord, v: any): void {
    if (null == v || true !== v.isVal || true === v.isTop || true === v.isNil ||
      rec.made.has(v.id) || rec.seen.has(v.id)) {
      return
    }

    // PART OF a written value is not a value beside it: report the
    // whole statement the author wrote, whichever piece of it the
    // fixpoint happened to meet here. See INNER_OF.
    let outer: any = v
    for (let up = this.containers.get(outer[INNER_OF]);
      null != up; up = this.containers.get(outer[INNER_OF])) {
      outer = up
    }
    if (outer !== v) {
      this.contribute(rec, outer)
      return
    }
    // Not the author's: see WRITTEN.
    if (true !== v[WRITTEN] && true !== v[FROM_SPREAD]) {
      return
    }
    if (true === v.isConjunct && Array.isArray(v.peg)) {
      rec.seen.add(v.id)
      for (const term of v.peg) {
        this.contribute(rec, term)
      }
      return
    }
    rec.seen.add(v.id)
    rec.conjuncts.push({
      canon: v.canon,
      id: v.id,
      role: roleOf(v),
      // COALESCED, unlike vet's siteOf: a `why` run reads whatever
      // source it was handed, and an inline document (a spec row, a
      // piped stdin) has no file name to stamp. The Go port answers
      // the empty string for the same value, so the two agree.
      site: {
        col: v.site.col, file: v.site.url ?? '', len: v.site.len,
        row: v.site.row,
      },
      src: v.site.src,
      ...(true === v.isPref ? { rank: v.rank } : {}),
    })
  }

  stands(path: string[], v: any): void {
    const key = path.join('.')
    const rec = this.paths.get(key)
    if (null != rec && 0 < rec.conjuncts.length) {
      return
    }
    this.record(path, v, undefined, undefined)
  }

  at(path: string[]): WhyConjunct[] {
    const rec = this.paths.get(path.join('.'))
    if (null == rec) {
      return []
    }
    const shown = new Map<string, Contribution>()
    const order = ['spread', 'ref', 'pref', 'literal']
    const out: Contribution[] = []
    for (const c of rec.conjuncts.slice().sort(cmpSite)) {
      if (c.site.row < 0) {
        out.push(c)
        continue
      }
      const key = [c.src, c.site.file,
        c.site.row, c.site.col, c.site.len].join('\u0000')
      const had = shown.get(key)
      if (null == had) {
        shown.set(key, c)
        out.push(c)
      }
      else if (order.indexOf(c.role) < order.indexOf(had.role)) {
        had.role = c.role
      }
    }
    return out.map(({ id, ...rest }) => rest)
  }
}
